### PR TITLE
Upgrade to Sentinel

### PR DESCRIPTION
**Pull request by BackEndTea**
This PR will replace Sentry with Sentinel.

* [x] Migrations for the Sentinel layout
* [x] Sentinel versions of our own interfaces (Authentication, Accountmanagement, etc.) 
* [x] Tests for the migrations and interfaces
* [x] Point the services to the new Sentinel versions

~~I'm not sure how to update the users passwords. Either every user needs to reset their password, which means that Sentry will be needed. Or this has to go into a major version up, with no backwards compatibility. So i'd like some input on that matter.~~

Follows #501
Related to #498

Besides a security audit, this is about finished. The only thing left is removing all the old Sentry classes and their tests.

### CODE REVIEW

**Comment by chartjes** [comment]
@BackEndTea How to get existing accounts to convert over. One suggestion to me was this:

* add a field to existing user accounts identifying them as "legacy"
* when a legacy account tried to log in we redirect them to a page where they can update their password for Sentinel (build something that sends them a token via email)
* update the field once people have added a new password

**Comment by BackEndTea** [comment]
I was surprised to find out that i was able to log in just fine with my old information when i changed between Sentinel and Sentry.
So after looking around a bit i found this:

The Sentry Hashing function:
```php 
	public function hash($string)
	{
		// Usually caused by an old PHP environment, see
		// https://github.com/cartalyst/sentry/issues/98#issuecomment-12974603
		// and https://github.com/ircmaxell/password_compat/issues/10
		if ( ! function_exists('password_hash'))
		{
			throw new \RuntimeException('The function password_hash() does not exist, your PHP environment is probably incompatible. Try running [vendor/ircmaxell/password-compat/version-test.php] to check compatibility or use an alternative hashing strategy.');
		}

		if (($hash = password_hash($string, PASSWORD_DEFAULT)) === false)
		{
			throw new \RuntimeException('Error generating hash from string, your PHP environment is probably incompatible. Try running [vendor/ircmaxell/password-compat/version-test.php] to check compatibility or use an alternative hashing strategy.');
		}

		return $hash;
	}
```

The Sentinel one
```php
    public function hash($value)
    {
        if (! $hash = password_hash($value, PASSWORD_DEFAULT)) {
            throw new RuntimeException('Error hashing value. Check system compatibility with password_hash().');
        }

        return $hash;
    }
```

At least for the native hashing. So if we use the native hashing functionality there is no need to add a 'legacy' field etc and make users reset their passwords. If we go that route this can basically be implemented already. I'd like to add something that destroys all sessions when the upgrade happens for good measure.
Though in that case i'd like someone else to take a look at this and test it out locally.

If someone is inclined to do so, i use this locally to change between Sentry and Sentinel on the fly
```php
$sentinel = true;
        if ($sentinel) {
            $app[Sentinel::class] = function () {
                return (new \Cartalyst\Sentinel\Native\Facades\Sentinel())->getSentinel();
            };

            $app[AccountManagement::class] = function ($app) {
                return new \OpenCFP\Infrastructure\Auth\SentinelAccountManagement($app[Sentinel::class]);
            };

            $app[IdentityProvider::class] = function ($app) {
                return new \OpenCFP\Infrastructure\Auth\SentinelIdentityProvider($app[Sentinel::class], new IlluminateSpeakerRepository(new User()));
            };

            $app[Authentication::class] = function ($app) {
                return new \OpenCFP\Infrastructure\Auth\SentinelAuthentication($app[Sentinel::class], $app[AccountManagement::class]);
            };
        } else {
            $app[AccountManagement::class] = function ($app) {
                return new SentryAccountManagement($app['sentry']);
            };

            $app[IdentityProvider::class] = function ($app) {
                return new SentryIdentityProvider($app['sentry'], new IlluminateSpeakerRepository(new User()));
            };

            $app[Authentication::class] = function ($app) {
                return new SentryAuthentication($app['sentry']);
            };
        }
```

**Comment by chartjes** [comment]
When I went back and looked at some of my scribbled notes when I was attempting to switch things over, I did see that in terms of hashing the password Sentinel was a drop-in replacement for Sentry. The changes were to any code that spoke to Sentry, and once we centralized all the auth/acl stuff I felt like we were in a much better place to do this.

I will ask for someone to do a security audit -- we had one maybe 18 months ago and it was very useful.

File: classes/Infrastructure/Auth/SentinelIdentityProvider.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+     * Retrieves the currently authenticate user's username.
+     *
+     * @throws NotAuthenticatedException
+     *
+     * @return User
+     */
+    public function getCurrentUser()
+    {
+        $user = $this->sentinel->getUser();
+        if ($user == null || !$user) {
```
Comment (kelunik) [inline classes/Infrastructure/Auth/SentinelIdentityProvider.php]: Just `if (!$user)` is equivalent.

Comment (BackEndTea) [reply classes/Infrastructure/Auth/SentinelIdentityProvider.php]: Thanks, i updated it to the following, making it a bit more accurate
```php
if (!$user instanceof \Cartalyst\Sentinel\Users\UserInterface) {
    throw new NotAuthenticatedException();
}
```

**Comment by chartjes** [comment]
There are currently two failing tests, which refer to tests for console actions, both of which are complaining about setting headers. I'm guessing the console commands need updating to use Sentinel

**Comment by BackEndTea** [comment]
It looks like it was an error with the tests Mocking Sentry hard coded instead of the AccountManagementInterface.

And while on the Subject of the tests for console actions, it may be smart to use the [Symfony KernelTestcase](https://symfony.com/doc/current/console.html#testing-commands) for those in the future. But thats outside of the scope for this PR

**Comment by chartjes** [comment]
Okay folks, I think it is time to decide on go vs no-go on this feature.

@BackEndTea if you can fix the conflicts it's complaining about here, I vote for pushing this feature. I am simply unsure if this qualifies as a v2 release or not

**Comment by mdwheele** [comment]
👍 

If you can checkout `master`, deploy the application, create a user and login, then merge `upgrade/sentinel`, run `script/update` and everything works okay, then it's probably okay to stay within 1.x. Since it looks like Sentinel is a drop-in for Sentry, then it looks like no password shenanigans need to be accounted for. 

The "public interface" that we version semantically seems to be "the experience for the CFP organizer". If that didn't change (they just upgrade as normal) then we're good to stay in a major.

*I think we should save 2.x for something super special and magical like #618* 😀

**Comment by chartjes** [comment]
After resolving the conflicts and merging into current master, I get the following errors in the tests

vendor/bin/phpunit --configuration phpunit.xml
PHPUnit 6.4.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.11 with Xdebug 2.5.5
Configuration: /Users/chartjes/Sites/opencfp/phpunit.xml

............................................................................................................... 111 / 416 ( 26%)
..........................................FFF........................SS........................................ 222 / 416 ( 53%)
............................................................................................................... 333 / 416 ( 80%)
...................................................................................                             416 / 416 (100%)

Time: 1.82 minutes, Memory: 318.00MB

There were 3 failures:

1) OpenCFP\Test\Integration\Infrastructure\Auth\SentinelUserTest::checkResetPasswordCodeWorks
Failed asserting that false is true.

/Users/chartjes/Sites/opencfp/tests/Integration/Infrastructure/Auth/SentinelUserTest.php:84

2) OpenCFP\Test\Integration\Infrastructure\Auth\SentinelUserTest::getResetPassWordCodeWorks
Failed asserting that false is true.

/Users/chartjes/Sites/opencfp/tests/Integration/Infrastructure/Auth/SentinelUserTest.php:100

3) OpenCFP\Test\Integration\Infrastructure\Auth\SentinelUserTest::attemptResetPasswordWorks
Failed asserting that false is true.

/Users/chartjes/Sites/opencfp/tests/Integration/Infrastructure/Auth/SentinelUserTest.php:122

--

There were 2 skipped tests:

1) OpenCFP\Test\Integration\MigrationsTest::oldUsersGetActivated

/Users/chartjes/Sites/opencfp/tests/Integration/MigrationsTest.php:45

2) OpenCFP\Test\Integration\MigrationsTest::adminsGetPromoted

/Users/chartjes/Sites/opencfp/tests/Integration/MigrationsTest.php:71

**Comment by chartjes** [comment]
But logging in did work! So I think once we get the test suite completely working we should be okay

**Comment by chartjes** [comment]
Some other things currently are not working:

* deleting talks
* logging out

It needs some more testing I think...

**Comment by BackEndTea** [comment]
I fixed those bugs, and upped unit test coverage to about as high as i can get it, without starting to write unit tests for a lot of things that are outside of the scope for this PR.

A lot of things are now tested in both unit and integration tests, but i prefer to keep them both. Since the integration tests will show us our errors in interfacing with Sentinel.

**Comment by chartjes** [comment]
I am *still* not able to get the tests to pass despite them passing on Travis. I wish I could figure out why.

composer install
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Nothing to install or update
Package guzzle/guzzle is abandoned, you should avoid using it. Use guzzlehttp/guzzle instead.
Generating autoload files
mysql -uroot -e "DROP DATABASE IF EXISTS cfp_test"
mysql -uroot -e "CREATE DATABASE cfp_test"
cp phinx.yml.dist phinx.yml
vendor/bin/phinx migrate -e testing
Phinx by CakePHP - https://phinx.org. 0.8.1

using config file ./phinx.yml
using config parser yaml
using migration paths
 - /Users/chartjes/Sites/opencfp/migrations
using environment testing
using adapter mysql
using database cfp_test

 == 0 Schema: migrating
 == 0 Schema: migrated 0.0579s

 == 20140823091041 Favorites: migrating
 == 20140823091041 Favorites: migrated 0.0091s

 == 20140828144530 AddIdToSpeakerTable: migrating
 == 20140828144530 AddIdToSpeakerTable: migrated 0.0006s

 == 20140828145029 AddHotelAndTransportationCostFields: migrating
 == 20140828145029 AddHotelAndTransportationCostFields: migrated 0.0284s

 == 20140902174929 RemoveSpeakerTable: migrating
 == 20140902174929 RemoveSpeakerTable: migrated 0.0372s

 == 20141121173412 AddTalkMeta: migrating
 == 20141121173412 AddTalkMeta: migrated 0.0097s

 == 20141122213222 CreateTalkComment: migrating
 == 20141122213222 CreateTalkComment: migrated 0.0124s

 == 20141201172522 AlterUserColumnDefaults: migrating
 == 20141201172522 AlterUserColumnDefaults: migrated 0.1267s

 == 20141203111136 AlterTalkColumnDefaults: migrating
 == 20141203111136 AlterTalkColumnDefaults: migrated 0.0449s

 == 20150518024706 CreateLeagueOauthTables: migrating
 == 20150518024706 CreateLeagueOauthTables: migrated 0.0948s

 == 20150519122926 ResetPhotoPaths: migrating
 == 20150519122926 ResetPhotoPaths: migrated 0.0021s

 == 20150702132854 FixDefaultsForUsersTable: migrating
 == 20150702132854 FixDefaultsForUsersTable: migrated 0.2084s

 == 20151012165609 AlterTalkColumnsAllowNull: migrating
 == 20151012165609 AlterTalkColumnsAllowNull: migrated 0.0243s

 == 20151214140200 Airports: migrating
 == 20151214140200 Airports: migrated 0.0326s

 == 20160106061843 FixZeroDates: migrating
 == 20160106061843 FixZeroDates: migrated 0.0711s

 == 20161224211743 TalksTagsMigration: migrating
 == 20161224211743 TalksTagsMigration: migrated 0.0156s

 == 20171019025925 AddNeedsProfileFlag: migrating
 == 20171019025925 AddNeedsProfileFlag: migrated 0.0145s

 == 20171107125309 ReviewerRoleMigration: migrating
 == 20171107125309 ReviewerRoleMigration: migrated 0.0011s

 == 20171120102354 SentinelMigration: migrating
 == 20171120102354 SentinelMigration: migrated 0.0844s

 == 20171120112704 ActivateUserMigration: migrating
 == 20171120112704 ActivateUserMigration: migrated 0.0270s

 == 20171120122725 RolesMigration: migrating
 == 20171120122725 RolesMigration: migrated 0.0022s

 == 20171120123411 UserRolesMigration: migrating
 == 20171120123411 UserRolesMigration: migrated 0.0101s

All Done. Took 0.9698s
vendor/bin/phpunit --configuration tests/Integration/phpunit.xml.dist
PHPUnit 6.4.4 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.1.11 with Xdebug 2.5.5
Configuration: /Users/chartjes/Sites/opencfp/tests/Integration/phpunit.xml.dist

............................................................................................................................................... 143 / 183 ( 78%)
...........FFF........................SS                                                                                                        183 / 183 (100%)

Time: 2 minutes, Memory: 316.00MB

There were 3 failures:

1) OpenCFP\Test\Integration\Infrastructure\Auth\SentinelUserTest::checkResetPasswordCodeWorks
Failed asserting that false is true.

/Users/chartjes/Sites/opencfp/tests/Integration/Infrastructure/Auth/SentinelUserTest.php:84

2) OpenCFP\Test\Integration\Infrastructure\Auth\SentinelUserTest::getResetPassWordCodeWorks
Failed asserting that false is true.

/Users/chartjes/Sites/opencfp/tests/Integration/Infrastructure/Auth/SentinelUserTest.php:100

3) OpenCFP\Test\Integration\Infrastructure\Auth\SentinelUserTest::attemptResetPasswordWorks
Failed asserting that false is true.

/Users/chartjes/Sites/opencfp/tests/Integration/Infrastructure/Auth/SentinelUserTest.php:122

--

There were 2 skipped tests:

1) OpenCFP\Test\Integration\MigrationsTest::oldUsersGetActivated

/Users/chartjes/Sites/opencfp/tests/Integration/MigrationsTest.php:45

2) OpenCFP\Test\Integration\MigrationsTest::adminsGetPromoted

/Users/chartjes/Sites/opencfp/tests/Integration/MigrationsTest.php:71

**Comment by BackEndTea** [comment]
@chartjes 
Its kinda hard debugging this for me. Could you run ```vendor/bin/phpunit --configuration tests/Integration/phpunit.xml.dist --stop-on-failure tests/Integration/Infrastructure/Auth/SentinelUserTest.php```
And check what the database looks like, especially the reminders and users table.

File: classes/Infrastructure/Auth/SentinelUser.php [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        return $this->sentinel->getReminderRepository()->exists($this->user, $resetCode) !== false;
+    }
+
+    /**
+     * Create a password reset code for the user, or reset it if it already exists
+     * This will NOT retrieve the code if it is already set, it will instead generate a new one and set that.
+     */
+    public function getResetPasswordCode(): string
+    {
+        return $this->sentinel->getReminderRepository()->create($this->user);
```
[the remark this answers is not in the document]
Comment (jorissteyn) [reply classes/Infrastructure/Auth/SentinelUser.php]: Looks good Gert! Tests run fine on my environment.

File: classes/Http/Controller/ForgotController.php [hunk 3]
```diff
@@ -90,7 +89,7 @@ public function resetAction(Request $req)
             if (!$user->checkResetPasswordCode($req->get('reset_code'))) {
                 ++$error;
```
[the remark this answers is not in the document]
Comment (BackEndTea) [reply classes/Http/Controller/ForgotController.php]: Thanks for the review, i fixed all the other errors, except for this one. Its outside the scope of this PR.

**Comment by BackEndTea** [comment]
The fix for being unable to delete talks is in a separate PR: #727

**Comment by mdwheele** [comment]
@BackEndTea resolved [the talk deletion bug](https://github.com/opencfp/opencfp/pull/727) and that has been merged.

**Comment by chartjes** [comment]
Thank you to EVERYONE for all their help with what is a major change of things behind the scenes.

**Comment by BackEndTea** [comment]
Thanks @chartjes for merging, and thanks to everyone for helping with this.

**Pull request merged**